src/hooks/useQueryParam/index.ts

```diff
@@ -0,0 +1,54 @@
+import { useCallback } from "react";
+import { useNavigate, useSearchParams } from "react-router-dom";
+import { parseQueryString, stringifyQuery } from "utils/query-string";
+
+/** `useQueryParams` returns all of the query params passed into the url */
+const useQueryParams = () => {
+  const [searchParams] = useSearchParams();
+  const navigate = useNavigate();
+  const setQueryString = useCallback(
+    (params: { [key: string]: any }) => {
+      const stringifiedQuery = stringifyQuery(params);
+      navigate(`?${stringifiedQuery}`, { replace: true });
+    },
+    [navigate]
+  );
+
+  return [parseQueryString(searchParams.toString()), setQueryString] as const;
+};
+
+/** `useQueryParam` allows you to interact with a query param in the same way you would use a useState hook.  */
+const useQueryParam = <T>(
```

Comment:
Were there any other particular use cases/types you had in mind when using a generic here?

Reply:
It just allows us to not have to type the type and get strong type safety.

It works in the same way as useState()

if you pass in a string it will ensure only strings are used.
If you pass in an array it will ensure only arrays are used.
number etc.

```ts
// str is of type string, setStr accepts only a string
const [str, setStr] = useQueryParam(queryParam, "something");
// num is of type number, setNum accepts only a number
const [num, setNum] = useQueryParam(queryParam, 1);
// it even allows you to assert a type if it can't be inferred
// For example if you initialize a null or empty array you can tell it that the array should be a number array
const [someArray, setSomeArray] = useQueryParam<number[]>(queryParam, [])
```